python/tvm/relay/transform/quantize/_average_max_channel_patterns.py

```diff
@@ -0,0 +1,205 @@
+    Conv2DPattern,
+    Conv2DBiasAddPattern,
+    DensePattern,
+    PerChannelPattern,
+    CalibrationCallback,
+    QuantizerPattern,
+    DatasetManager,
+)
+
+
+class AverageMaxPerChannelConv2DPattern(Conv2DPattern, PerChannelPattern):
+    """Per channel version of Conv2DPattern, implementing the average max algorithm to
+    calculate scales and zero points."""
+
+    def __init__(self, calibration_callback: CalibrationCallback = None):
+        super().__init__(calibration_callback)
+
+    def extract_attrs(self, pre, post, node_map):
+        conv2d = node_map[self.conv2d][0]
+        weight = node_map[self.conv_weight][0]
+
+        self.get_attrs(conv2d.attrs, weight.checked_type.shape)
+        return post
+
+    def scale(self, name, is_weight=False):
+        if is_weight:
+            shape = (self.channels,)
+        else:
+            shape = ()
+        var = relay.var(
+            str(name) + "_scale_" + str(QuantizerPattern.scales_count), shape=shape, dtype="float32"
+        )
+        QuantizerPattern.scales_count += 1
+        return var
+
+    def calibrate_pattern(self, calibration_info):
+        self.attr_callback(calibration_info.partition_info.expr)
+        scale_zp_values = {}
+
+        data_min_sum = 0
+        data_max_sum = 0
+
+        weight_min_sums = np.zeros(shape=(self.channels,))
+        weight_max_sums = np.zeros(shape=(self.channels,))
+
+        while not calibration_info.dataset_manager.is_empty():
+            # Get the original input from dataset manger, run unquantized graph with those inputs
+            image_list, _ = calibration_info.dataset_manager.get_next_batch()
+            unquantized_inputs = calibration_info.get_unquantized_layer_inputs(image_list)
+
+            data = unquantized_inputs[0]
+            weight = unquantized_inputs[1]
+
+            data_min_sum += np.min(data)
+            data_max_sum += np.max(data)
+
+            weight_min_sums += np.min(weight, axis=list(range(len(weight.shape))).remove(0))
+            weight_max_sums += np.max(weight, axis=list(range(len(weight.shape))).remove(0))
+
+        calibration_info.dataset_manager.reset()
+
+        data_min_avg = data_min_sum / calibration_info.dataset_manager.num_batches()
+        data_max_avg = data_max_sum / calibration_info.dataset_manager.num_batches()
+
+        weight_min_avgs = weight_min_sums / calibration_info.dataset_manager.num_batches()
+        weight_max_avgs = weight_max_sums / calibration_info.dataset_manager.num_batches()
+
+        # Threshold for quantization of an input to a layer is mean(abs(avg_max), abs(avg_min))
+        data_threshold = (np.abs(data_min_avg) + np.abs(data_max_avg)) / 2
+        weight_thresholds = (np.abs(weight_min_avgs) + np.abs(weight_max_avgs)) / 2
+
+        # Since this is a symmetric distribution and we are quantizing to int8, there are 256 bins, and 128 are positive
+        data_scale = data_threshold / 128
+        weight_scales = weight_thresholds / 128
```

Comment:
There is a lot of duplication in this. Perhaps we could 
1) make data_min_max_sums a numpy array of size 2, put the min int he first slot, and the max in the second, and then use np.avg(np.abs(x))/128 to get scales?
2) make this a utility function to reduce the copied code?

Reply:
The reason I didn't do the numpy array in this case is because this is the per channel version. I was having trouble getting it to work because I need to sum across an axis for weight (and average across it as well). I'm not super great at making numpy do what I want though so if you know a way of dealing with the axis I'm happy to do it

Also I know it is long, but I do think that leaving it longer format makes it clearer what's happening with the per channel calculation